keps/sig-storage/1472/README.md

```diff
@@ -0,0 +1,1416 @@
+difficult. If the capacity represents "maximum volume size 10GiB", it may be possible
+to create exactly one such volume or several, so rejecting the
+pool after one volume could be a false negative. With "available
+capacity 10GiB" it may or may not be possible to create two volumes of
+5GiB each, so accepting the node for two such volumes could be a false
+positive.
+
+More promising might be to add prioritization of nodes based on how
+much capacity they have left, thus spreading out storage usage evenly.
+This is a likely future extension of this KEP.
+
+Either way, the problem of recovering more gracefully from running out
+of storage after a bad scheduling decision will have to be addressed
+eventually. Details for that are in https://github.com/kubernetes/enhancements/pull/1703.
+
+### Test Plan
+
+The Kubernetes scheduler extension will be tested with new unit tests
+that simulate a variety of scenarios:
+- different volume sizes and types
+- driver with and without storage capacity tracking enabled
+- capacity information for node local storage (node selector with one
+  host name), network attached storage (more complex node selector),
+  storage available in the entire cluster (no node restriction)
+- no suitable node, one suitable node, several suitable nodes
+
+Producing capacity information in external-provisioner also can be
+tested with new unit tests. This has to cover:
+- different modes
+- different storage classes
+- a driver response where storage classes matter and where they
+  don't matter
+- different topologies
+- various older capacity information, including:
+  - no entries
+  - obsolete entries
+  - entries that need to be updated
+  - entries that can be left unchanged
+
+This needs to run with mocked CSI driver and API server interfaces to
+provide the input and capture the output.
+
+Full end-to-end testing is needed to ensure that new RBAC rules are
+identified and documented properly. For this, a new alpha deployment
+in csi-driver-host-path is needed because we have to make changes to
+the deployment like setting `CSIDriver.spec.storageCapacity` which
+will only be valid when tested with Kubernetes clusters where
+alpha features are enabled.
+
+The CSI hostpath driver needs to be changed such that it reports the
+remaining capacity of the filesystem where it creates volumes. The
+existing raw block volume tests then can be used to ensure that pod
+scheduling works:
+- Those volumes have a size set.
+- Late binding is enabled for the CSI hostpath driver.
+
+A new test can be written which checks for `CSIStorageCapacity` objects,
+asks for pod scheduling with a volume that is too large, and then
+checks for events that describe the problem.
+
+### Graduation Criteria
+
+#### Alpha -> Beta Graduation
+
+- Gather feedback from developers and users
+- Integration with [Cluster Autoscaler](https://github.com/kubernetes/autoscaler)
+- Generic solution for identifying storage pools
```

Comment:
Can you reference this KEP here (https://github.com/kubernetes/enhancements/pull/1347)?

Also add that we need to figure out how to do spreading.  Maybe it is covered by storage pools but it should be called out.

Reply:
Done.